src/lib/structures/modelMap.ml

```diff
@@ -151,22 +176,37 @@ let rec subst_in_term id e c =
 let subst id e { values; suspicious } =
   if not @@ Expr.is_model_term e then
     raise (Errors.Error (Model_error (Subst_not_model_term e)));
-  let values = P.map (Graph.map (subst_in_term id e)) values in
+
+  let values =
+    P.map
+      (fun graph ->
+         match graph with
+         | C constraints -> C (Constraints.map (subst_in_term id e) constraints)
+         | Free a -> Free (subst_in_term id e a)
```

Comment:
It was my first code but let's imagine you're using the library and you get a model with `SAT.get_model`. One of the value of the model is an abstract value. You may want to substitute it by something else. Sometimes, it will work because this is not a free graph, sometimes the function `subst` silently ignores your request. 

My code isn't good because after doing this substitution, the user could add a new constraint to the graph but the current `add` function will erase the effect of the previous substitution.

Reply:
The model is an output value of the solver, I don't think the use case where we allow people to alter the model after the fact is one that we want to support. If people want to change the model they need to add assertions to the problem and call `solve` again to get a new model.

Under that assumption, the semantics of the `Free` constructor is "this function is unconstrained", and the inner abstract value is a "dummy" name that is guaranteed to not occur anywhere else in the model (by construction). Given that semantics, we shouldn't substitute under the `Free` constructor.

If we want to have a proper semantics for the graphs, the only easy way I see given the current architecture is a record with a default value instead, and print something like `(define-fun f (x Int) (ite (= x 0) 0 @a0))` when then only constraint is `[0] -> 0` — but again, I don't think we need to do this.

Reply:
> Under that assumption, the semantics of the `Free` constructor is "this function is unconstrained", and the inner abstract value is a "dummy" name that is guaranteed to not occur anywhere else in the model (by construction). Given that semantics, we shouldn't substitute under the `Free` constructor.

Note, I don't care either way — I think both implementations are OK as long as we don't expect the user to modify the model. If we do, neither implementation is correct as you point out (but again — I don't think we want to support this).

Reply:
I agree we don't need to support it. At some point, we should stop to expose `subst` in the library.